src/state-representation.lisp

```diff
@@ -0,0 +1,277 @@
+    :documentation  "A reference to the original pointer of amplitude memory, so the amplitudes can sit in the right place at the end of a computation."))
+  (:documentation "A PURE-STATE contains the quantum mechanical state for a system that can be described by a vector |ψ> of N qubits with unit length in a Hilbert space. The elements of this length 2^N vector is represented by AMPLITUDES."))
+
+(defun make-pure-state (num-qubits &key (allocation nil))
+  "ALLOCATION is an optional argument with the following behavior.
+    - If NULL (default), then a standard wavefunction in the Lisp heap will be allocated.
+    - If STRING, then the wavefunction will be allocated as a shared memory object, accessible by that name.
+    - Otherwise, it's assumed to be an object that is compatible with the ALLOCATION-LENGTH and ALLOCATE-VECTOR methods.
+    - will probs have to redo this in multiple places, have a helper function do the allocation stuff."
+  (let ((allocation
+          (etypecase allocation
+            (null
+             (make-instance 'lisp-allocation :length (expt 2 num-qubits)))
+            (string
+             (make-instance 'posix-shared-memory-allocation :length (expt 2 num-qubits)
+                                                            :name allocation))
+            (t
+             (assert (= (allocation-length allocation) (expt 2 num-qubits)))
+             allocation))))
+    (multiple-value-bind (amplitudes finalizer)
+        (allocate-vector allocation)
+      ;; initialize amplitudes to zero state (efficiently)
+      (setf (aref amplitudes 0) (cflonum 1)) 
+      (let ((state (make-instance 'pure-state :num-qubits num-qubits 
+                                              :amplitudes amplitudes 
+                                              :allocation allocation)))
+        ;; When the state disappears, make sure the shared
+        ;; memory gets deallocated too.
+        (tg:finalize state finalizer)
+        state))))
+
+(defmethod initialize-instance :after ((state pure-state) &key num-qubits &allow-other-keys)
+  ;; If AMPLITUDES are provided, assert that the length of the vector
+  ;; is representative of NUM-QUBITS qubits. If AMPLITUDES are not
+  ;; provided, create a vector of AMPLITUDES for NUM-QUBITS
+  ;; qubits. Also, create a TRIAL-AMPLITUDES vector and save a
+  ;; reference to the originally provided memory in AMPLITUDES.
+  (cond 
+    ((and (slot-boundp state 'amplitudes)
+          (not (null (slot-value state 'amplitudes))))
+     (assert (<= num-qubits (wavefunction-qubits (amplitudes state)))
+             ()
+             state
+             (wavefunction-qubits (amplitudes state))
+             num-qubits))
+    (t
+     (setf
+      ;; Initialize the AMPLITUDES and TRIAL-AMPLITUDES to an empty
+      ;; array of the correct size.
+      (amplitudes state) (make-lisp-cflonum-vector (expt 2 num-qubits)))))
+  (setf
+     (%trial-amplitudes state) (make-lisp-cflonum-vector (expt 2 num-qubits))
```

Comment:
That sounds reasonable. I think a good way to do this would be to test for the presence of superoperator-defintions or any noise values in the `initialize-instance` methods of the different types of qvms with pure state representations, then allocate if found.

Reply:
Also, sorry -- I'm confused about your first comment. The allocation should only be enough space for `amplitudes`, so shouldn't I have to allocate additional space for `trial-amplitudes`, instead of using the same allocation?

Reply:
Oops, sorry I could have chosen my words better. By `allocation`, I meant the `quantum-system-state` slot by that name, not the actual memory that gets allocated. The `allocation` slot holds an "allocation description" which describes the type of allocation, i.e. `lisp-allocation`, `posix-shared-memory-allocation`, or `c-allocation`, plus the size of the allocation. The actual allocating of memory is then done by passing that allocation description object to `allocate-vector`, which takes care of actually allocating the memory in the right way/place.

So what I meant is that since we store the allocation description, potentially `%trail-amplitudes` could be allocated in the same way as the `amplitudes` vector. Currently, `%trail-amplitudes` is always allocated by calling `make-lisp-cflonum-vector`, which is equivalent (I think could be missing details) to calling `allocate-vector` and passing it a `lisp-allocation` description.  But if the user request `c-allocation`, say, then `amplitudes` would be allocated as unmanaged foreign memory, whereas `%trail-amplitudes` is still allocated on the lisp heap.

Again, I think this is the pre-existing behavior inherited from the old noisy-qvm, so maybe it's undesirable or was deemed unnecessary to have `%trail-amplitudes` be allocated in the same way as `amplitudes`. Worth asking the experts aka @stylewarning if not already trapped on a 37-hour flight. Or maybe there's wifi onboard and he can beam QVM knowledge to us from 30,000 ft. ✈️

Reply:
gotcha, thanks for clearing that up :)

Reply:
I think we could consider the `allocation` of trial-amplitudes in a follow up PR. In this PR I implemented your second suggestion.

Reply:
works for me